### PR TITLE
Update Alertmanager in Mimir

### PR DESCRIPTION
**Pull request by grobinson-grafana**
#### What this PR does

This pull request updates Alertmanager in Mimir. It deprecates Alertmanager APIv1. The version of Alertmanager vendored in this commit contains support for UTF-8, however is disabled until we initialize the compat package from within Mimir using `compat.InitFromFlags`. UTF-8 will be enabled at a later time.

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by grobinson-grafana** [comment]
Waiting for https://github.com/prometheus/alertmanager/pull/3668.

**Comment by grobinson-grafana** [comment]
I tested this commit in `dev-us-central-0` using commit [`3537d35`](https://github.com/grafana/backend-enterprise/commit/3537d353ca865c5b4b362caa5471102b5a12e2b5). You can see from the link that the commit from Mimir is `5a22d1c`.

Here is a screenshot showing the version on startup:

Screenshot 2024-01-18 at 4 27 13 PM

### Test 1: mimirtool works as expected with non UTF-8 configurations

`mimirtool` can load the following configuration:

```yml
route:
    receiver: default
    routes:
        - continue: false
          matchers:
            - foo=bar
            - bar=baz
          mute_time_intervals: []
          receiver: ""
          routes: []
```

```
mimirtool alertmanager load alertmanager.yml
```

and the configuration can be seen in Grafana:

Screenshot 2024-01-18 at 4 30 23 PM

### Test 2: UTF-8 configurations are rejected

Using a modified version of mimirtool that supports UTF-8, it cannot load UTF-8 configurations as the request is rejected in the API:

```yml
route:
    receiver: default
    routes:
        - continue: false
          matchers:
            - foo=bar
            - bar=baz
            - baz="\xf0\x9f\x99\x82"
            - bar🙂=baz
          mute_time_intervals: []
          receiver: ""
          routes: []
```

```
mimirtool alertmanager load alertmanager.yml
ERRO[0000] response                                      body="error validating Alertmanager config: bad matcher format: bar🙂=baz\n" status="400 Bad Request"
mimirtool: error: POST request to https://alertmanager-dev-us-central1.grafana-dev.net/api/v1/alerts failed: server returned HTTP status: 400 Bad Request, body: "error validating Alertmanager config: bad matcher format: bar🙂=baz\n", try --help
```

The input `baz="\xf0\x9f\x99\x82"` is accepted, but `bar🙂=baz` is not. The reason for this is that in current versions of Alertmanager `"\xf0\x9f\x99\x82"` is interpreted as a string literal, and the escape sequences are not interpreted. Once UTF-8 is enabled this will change such that `"\xf0\x9f\x99\x82"` will be interpreted as  `🙂`. These examples are known and are referred to as disagreement. We are planning to log all occurrences of disagreement, and fix them if required, in each cell before enabling UTF-8.

Screenshot 2024-01-18 at 4 37 18 PM

Screenshot 2024-01-18 at 4 37 30 PM

### Test 3: amtool works as expected with non UTF-8 alerts and silences

amtool can create alerts and silences without UTF-8 as normal:

```
amtool alert add foo=bar
Warning: amtool version (0.26.0) and alertmanager version (grobinson-update-am-3537d353) are different.
```

Screenshot 2024-01-18 at 4 42 33 PM

```
amtool silence add foo=bar -c "Silence foo=bar"
Warning: amtool version (0.26.0) and alertmanager version (grobinson-update-am-3537d353) are different.
6c386bdf-d1ad-4c83-a23d-5e12d1bce0bb
```

Screenshot 2024-01-18 at 4 43 05 PM

### Test 4: UTF-8 alerts and silences are rejected 

A modified version of amtool that supports UTF-8 cannot create alerts or silences containing UTF-8 as these are rejected in the API:

```
amtool alert add foo🙂=bar
Warning: amtool version (0.26.0) and alertmanager version (grobinson-update-am-3537d353) are different.
amtool: error: [POST /alerts][400] postAlertsBadRequest  invalid label set: invalid name "foo🙂"
```

```
amtool silence add foo🙂=bar -c "Silence foo🙂=bar"
Warning: amtool version (0.26.0) and alertmanager version (grobinson-update-am-3537d353) are different.
amtool: error: [POST /silences][400] postSilencesBadRequest  silence invalid: invalid label matcher 0: invalid label name "foo🙂"
```

### Test 5: UTF-8 is allowed on the right hand side 

Of course, UTF-8 is allowed on the right hand side (either as the value of a label or right hand side of a matcher). For example:

```yml
route:
    receiver: default
    routes:
        - continue: false
          matchers:
            - foo=🙂bar
            - bar=🙂baz
          mute_time_intervals: []
          receiver: ""
          routes: []
```

Screenshot 2024-01-18 at 4 47 30 PM

```
amtool alert add foo=🙂bar
Warning: amtool version (0.26.0) and alertmanager version (grobinson-update-am-3537d353) are different.
```

Screenshot 2024-01-18 at 4 48 08 PM

```
amtool silence add foo=🙂bar -c "Silence foo=🙂bar"
Warning: amtool version (0.26.0) and alertmanager version (grobinson-update-am-3537d353) are different.
bc5483b2-69a6-430e-8506-b52f3fa95104
```

Screenshot 2024-01-18 at 4 48 41 PM

**Comment by grobinson-grafana** [comment]
### Test 6: APIv1 has been deprecated

Here is a test showing that APIv1 has been deprecated in `grobinson-update-am-3537d353`:

```
curl https://alertmanager-dev-us-central1.grafana-dev.net/alertmanager/api/v1/alerts
{"status":"deprecated","error":"The Alertmanager v1 API was deprecated in version 0.16.0 and is removed as of version 0.28.0 - please use the equivalent route in the v2 API"}
```

and here is `grafana/metrics-enterprise:r273-0f9172ac`:

```
curl https://alertmanager-dev-us-central1.grafana-dev.net/alertmanager/api/v1/alerts
{"status":"success","data":[]}
```

**Review by pracucci — COMMENTED** [review]
Could you add a CHANGELOG entry to describe the changes (including the removal of API v1), please?

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 ### Grafana Mimir
 
+* [CHANGE] Alertmanager: Updated Alertmanager to latest main (`fa6a7e6`). This version of Alertmanager adds support for UTF-8 but is disabled in Mimir. It also deprecates Alertmanager APIv1, and replaces all APIv1 endpoints with a deprecation notice.
```
Comment (pracucci) [inline CHANGELOG.md:7]: [nit] A couple of nits:

1. In the CHANGELOG we only report information useful for the final user (e.g. no commit hashes or no reference to features not enabled in Mimir)
2. We add the PR number at the end of the CHANGELOG entry

So, something like this (but feel free to edit it):

```suggestion
* [CHANGE] Alertmanager: Alertmanager API v1 endpoints support have been removed. The API v1 endpoints will not work and will only report a deprecation notice. #7103
```

Comment (pracucci) [reply CHANGELOG.md:7]: But what's about `/v1/silence` and `/v1/silences`? Are they still supported? Maybe it's better if you list all the endpoints whose support has been removed.

**Review by pracucci — COMMENTED** [review]
I went very quickly through the changes and I don't see anything major. I let @gotjosh do the proper review.

File: CHANGELOG.md (hunk 1 above)
Comment (grobinson-grafana) [reply CHANGELOG.md:7]: All APIv1 endpoints are deprecated upstream. I will make this clear in the CHANGELOG. Thanks for the feedback 👍

**Comment by grobinson-grafana** [comment]
Updated Alertmanager to commit [`f92a08d`](https://github.com/prometheus/alertmanager/commit/f92a08d073863b7e3731e503c45c81065c4b2910) include https://github.com/prometheus/alertmanager/pull/3676 that removes some unused code.

**Comment by grobinson-grafana** [comment]
@gotjosh something I thought about is whether we should also add this change in this PR instead of #6898:

```go
diff --git a/vendor/github.com/grafana/mimir/pkg/mimir/modules.go b/vendor/github.com/grafana/mimir/pkg/mimir/modules.go
index 58a03273e..9cc23ae3c 100644
--- a/vendor/github.com/grafana/mimir/pkg/mimir/modules.go
+++ b/vendor/github.com/grafana/mimir/pkg/mimir/modules.go
@@ -26,6 +26,8 @@ import (
        "github.com/opentracing-contrib/go-stdlib/nethttp"
        "github.com/opentracing/opentracing-go"
        "github.com/pkg/errors"
+       "github.com/prometheus/alertmanager/featurecontrol"
+       "github.com/prometheus/alertmanager/matchers/compat"
        "github.com/prometheus/client_golang/prometheus"
        "github.com/prometheus/common/config"
        "github.com/prometheus/prometheus/model/labels"
@@ -881,6 +883,10 @@ func (t *Mimir) initRuler() (serv services.Service, err error) {
 }

 func (t *Mimir) initAlertManager() (serv services.Service, err error) {
+       f, err := featurecontrol.NewFlags(util_log.Logger, featurecontrol.FeatureClassicMode)
+       util_log.CheckFatal("initializing Alertmanager feature flags", err)
+       compat.InitFromFlags(util_log.Logger, compat.RegisteredMetrics, f)
+
        t.Cfg.Alertmanager.ShardingRing.Common.ListenPort = t.Cfg.Server.GRPCListenPort
        t.Cfg.Alertmanager.CheckExternalURL(t.Cfg.API.AlertmanagerHTTPPrefix, util_log.Logger)
```

There will be no change in behavior. The difference is that we initialize the compat package with a logger, so we will get debug level logs. The [default variables](https://github.com/prometheus/alertmanager/blob/main/matchers/compat/parse.go#L32-L36) use `log.NewNopLogger()`.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 ### Grafana Mimir
 
+* [CHANGE] Alertmanager: Updated to latest main. It deprecates Alertmanager APIv1, and replaces all APIv1 endpoints with a deprecation notice. You should replace all use of the /v1/alerts, /v1/silences, /v1/receivers and /v1/status endpoints with their APIv2 equivalent.
```
Comment (gotjosh) [inline CHANGELOG.md:7]: Ditto from marco's suggestion.

- We don't use wording such as "Updated to latest main"; we simply list what the changes do for the end user.
- I'd recommend a list the endpoints directly.

```suggestion
* [CHANGE] Alertmanager: Deprecates the Alertmanager `v1` API. All `v1` API endpoints now respond with a JSON deprecation notice and a status code of `410`. All endpoints have a `v2` equivalent. The list of endpoints is:
  * `<alertmanager-web.external-url>/api/v1/alerts`
  * `<alertmanager-web.external-url>/api/v1/silence`
  * `<alertmanager-web.external-url>/api/v1/silences`
  * `<alertmanager-web.external-url>/api/v1/status`
  * `<alertmanager-web.external-url>/api/v1/receivers`
  * `<alertmanager-web.external-url>/api/v1/alerts`
```

**Review by gotjosh — COMMENTED** [review]
Overall LGTM, but please see my comments.

File: vendor/github.com/prometheus/alertmanager/matchers/compat/metrics.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+var RegisteredMetrics = NewMetrics(prometheus.DefaultRegisterer)
+
+type Metrics struct {
+	Total             *prometheus.GaugeVec
+	DisagreeTotal     *prometheus.GaugeVec
+	IncompatibleTotal *prometheus.GaugeVec
+	InvalidTotal      *prometheus.GaugeVec
+}
+
+func NewMetrics(r prometheus.Registerer) *Metrics {
```
[the remark this answers is not in the document]
Comment (grobinson-grafana) [reply vendor/github.com/prometheus/alertmanager/matchers/compat/metrics.go:40]: Those were available in dev when I tested just from the default registrar. Should we use another registrar? If so, we can initialize the compat package https://github.com/grafana/mimir/pull/7103#issuecomment-1900501338.

File: pkg/alertmanager/distributor_test.go [hunk 4]
```diff
@@ -84,16 +84,6 @@ func TestDistributor_DistributeRequest(t *testing.T) {
 			expStatusCode:      http.StatusOK,
 			expectedTotalCalls: 3,
 			route:              "/alerts",
```
[the remark this answers is not in the document]
Comment (grobinson-grafana) [reply pkg/alertmanager/distributor_test.go:86]: I don't think these tests call the Alertmanager API, as the response is stubbed in the test. Instead, the tests assert that the distributor fans out the request and merges the responses. I think I'll need to write some new integration tests for this.

File: integration/alertmanager_test.go
[the remark this answers is not in the document]
Comment (grobinson-grafana) [reply integration/alertmanager_test.go:1]: 1. Removed the comment 🙂 
2. This is checking the Mimir configuration API that returns YAML, so I believe this is correct. If I change it to `v2` then it 404s.

File: pkg/alertmanager/distributor_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/alertmanager/distributor_test.go:86]: I missed that - if the response is stubbed then I don't think we need to do anything to these tests. Instead, I suggest you add the cases in `integration/alertmanager_test.go`.

Replacing the removed cases for GET /v1/silences, GET /v1/silence/{id} and GET /v1/alerts with a simple call that asserts on the removal.

[the remark this answers is not in the document]
Comment (grobinson-grafana) [reply pkg/alertmanager/distributor_test.go:86]: Just pushed those tests 😄

File: vendor/github.com/prometheus/alertmanager/matchers/compat/metrics.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (grobinson-grafana) [reply vendor/github.com/prometheus/alertmanager/matchers/compat/metrics.go:40]: It seems Mimir uses the default registerer. If I attempt to re-register the metrics using the registerer in `initAlertManager` then I get a `panic: duplicate metrics collector registration attempted`.

**Review by gotjosh — APPROVED** [review]
LGTM

**Pull request merged**